rfcs/012-session-persistence/012-session-persistence.md

```diff
@@ -0,0 +1,332 @@
+  hibernating --> |error| error
+  removing --> deleted
+  deleted --> no-session
+  error --> no-session
+```
+
+### Launch Flow
+
+With the proposed solution, the decision tree to launch a session is similar to what is depicted as follows. Note that
+in all possible states, the user is informed about the current state and can make a clear decision on how to proceed
+when needed.
+
+```mermaid
+graph TD
+  launch[Session launch] 
+  is-running{is session running?}
+  pv-exists{session PV exists}
+  launch --> pv-exists
+  pv-exists --> |yes| head{repo at branch head?}
+  connect(connect to session)
+  
+  %% repo at HEAD, check if session is running
+  head --> |yes| is-running
+  
+  %% repo _not_ at HEAD, check the state 
+  head --> |no| gitstate{repo is...}
+
+  %% repo behind HEAD, check if ff-possible
+  gitstate --> |behind, FF possible| check-ff(Message 1)
+
+  %% FF possible, do it and check if session is running
+  check-ff --> |fast forward| ff(fast-forward & restart)
+  ff --> start
+  %% Done - connected to existing session, repo fast-forwarded
+  is-running --> |yes| connect
+  %% Done - started to new session, repo fast-forwarded
+  is-running --> |no| start(start a new session)
+
+  %% Cannot FF - either connect & fix or discard & launch new (at HEAD)
+  check-ff --> |connect| is-running
+  connect-discard --> |connect| is-running
+  connect-discard --> |discard| discard-restart
+
+  %% Repo ahead, dirty or "weird" - connect & fix or discard
+  gitstate --> |ahead or dirty HEAD| dirty(Message 2)
+  dirty --> |connect| is-running
+  dirty --> |discard| discard-restart(Discard PV)
+  discard-restart --> start
+  gitstate --> |weird| connect-discard(Message 3)
+  
+  check-ff --> |discard| discard-restart
+  
+  %% No PV means also no session - start a new one at HEAD
+  pv-exists --> |no| start
+  
+  start --> connect
+```
+
+`Message 1`, `Message 2`, and `Message 3` in the above diagram are displayed to the users when a decision should be
+made. We possibly show a diff of the persisted state and the requested state to the user and ask them to choose one the
+several possible paths to proceed to the launch. These messages along with possible followup paths are similar to the
+following list:
+
+- Message 1: You launched a session earlier on this project but the content seems older
+
+  - `Fast forward to the latest version` picking the last commit on `<branch>`
+  - `Connect to the existing session`
+  - `Discard the changes` and pick the last commit on `<branch>`
+
+- Message 2: You launched a session earlier on this project, and it contains new unsaved work
+
+  - `Connect to the session` and continue from there
+  - `Discard the changes` and pick the last commit on `<branch>`
+
+- Message 3: You launched a session earlier on this project and the state has diverged from the remote repository
+
+  - `Connect to the session` and continue from there
+  - `Discard the changes` and pick the last commit on `<branch>`
```

Comment:
I would strongly suggest to drop the logic around `git` state entirely from this RFC.

As mentioned above:

> From the user's perspective, connecting to a session, quitting, and reconnecting again is like coming and going from a **workspace** - they don't think about it in terms of the underlying git repository state, and we should not expect them to.

Users should be offered two choices:
1. Connect to the running or hibernating session (if possible indicate when was the last time they worked on it).
2. Discard the session and lose any unsaved work.

This creates a simple mental model around sessions for users and also reduces the implementation effort compared to the current proposal.

Reply:
I think it's ok to allow some notion of git in the RFC. Renku users are familiar with git anyhow and having some simple operations (e.g. fast-forwarding) should be easily understandable by them. In some places in the RFC, we talk about showing a diff of changes to users (as a nice-to-have), so, we have references to git anyway. Let's discuss this in a review meeting later.

Reply:
I think this has been addressed in the implementation.